doc/SONiC_OC_NTP_HLD.md

```diff
@@ -0,0 +1,491 @@
+- trusted key   
+- broadcast mode   
+
+### 1.1.6 Scalability Requirements
+
+### 1.1.7 Warm Boot Requirements
+
+## 1.2 Design Overview
+
+### 1.2.1 Basic Approach
+Implement NTP support using transformer in sonic-mgmt-framework.
+
+### 1.2.2 Container
+The front end code change will be done in management-framework container including:   
+- XML file for the CLI   
+- Python script to handle CLI request (actioner)   
+- Jinja template to render CLI output (renderer)   
+- OpenConfig YANG model for NTP openconfig-system.yang and openconfig-system-ext.yang   
+- SONiC NTP model for NTP based on Redis DB schema of NTP   
+- transformer functions to    
+   * convert OpenConfig YANG model to SONiC YANG model for NTP related configurations   
+   * convert from Linux command "ntpq -p" output to OpenConfig NTP state YANG model   
+
+### 1.2.3 SAI Overview
+
+# 2 Functionality
+
+## 2.1 Target Deployment Use Cases
+Manage/configure Management VRF via gNMI, REST and CLI interfaces
+
+## 2.2 Functional Description
+Provide CLI, gNMI and REST supports for Management VRF handling
+
+## 2.3 Backend change to support new configurations
+Provide change in hostcfgd, ntp config script, ntp service script.
+SONiC click CLI enhancement if possible.
+
+# 3 Design
+
+## 3.1 Overview
+
+Enhancing the management framework backend code and transformer methods to add support for NTP.
+
+## 3.2 DB Changes
+
+### 3.2.1 CONFIG DB
+This feature will allow the user to make NTP configuration changes to CONFIG DB, and get NTP peer states.
+
+### 3.2.2 APP DB
+
+### 3.2.3 STATE DB
+
+### 3.2.4 ASIC DB
+
+### 3.2.5 COUNTER DB
+
+## 3.3 Switch State Service Design
+
+### 3.3.1 Orchestration Agent
+
+### 3.3.2 Other Process
+
+## 3.4 SyncD
+
+## 3.5 SAI
+
+## 3.6 User Interface
+
+### 3.6.1 Data Models
+
+YANG models needed for NTP handling in the management framework:
+1. **openconfig-system.yang**  
+
+2. **openconfig-system-ext.yang**
+ 
+3. **sonic-system-ntp.yang**
+
+Supported yang objects and attributes:
+```diff
+
+module: openconfig-system 
+      +--rw system
+
++     +--rw ntp
++     |  +--rw config
+      |  |  +--rw enabled?                           boolean
++     |  |  +--rw ntp-source-address?                oc-inet:ip-address
+      |  |  +--rw enable-ntp-auth?                   boolean
++     |  |  +--rw oc-sys-ext:ntp-source-interface
++     |  |  |  +--rw oc-sys-ext:interface?      -> /oc-if:interfaces/interface/name
++     |  |  |  +--rw oc-sys-ext:subinterface?   -> /oc-if:interfaces/interface[oc-if:name=current()/../interface]/subinterfaces/subinterface/index
++     |  |  +--rw oc-sys-ext:vrf?                    string
+      |  +--ro state
+      |  |  +--ro enabled?                           boolean
++     |  |  +--ro ntp-source-address?                oc-inet:ip-address
+      |  |  +--ro enable-ntp-auth?                   boolean
+      |  |  +--ro auth-mismatch?                     oc-yang:counter64
++     |  |  +--ro oc-sys-ext:ntp-source-interface

```

Comment:
Can we configure both source-address and source-interface? .

Reply:
From REST/gNMI, yes, either source-address or source-interface. The last configured one will be used.

Reply:
what is the meaning of "the last configured one will be used" Please describe, as this might be equivalent to source-address itself, so would not get any perceived benefit from the addition of source-interface (which was ostensibly to not have to configure a fixed IP address in the configuration).